go/pkg/cas/upload.go

```diff
@@ -163,30 +171,51 @@ func (u *uploader) startProcessing(ctx context.Context, in *UploadInput) error {
 		// Do not use os.Stat() here. We want to know if it is a symlink.
 		info, err := os.Lstat(absPath)
 		if err != nil {
-			return errors.Wrapf(err, "lstat failed")
+			return errors.WithStack(err)
 		}
 
-		_, err = u.visitFile(ctx, absPath, info)
+		key := fsCacheKey{
+			AbsPath:            absPath,
+			PreserveSymlinks:   in.PreserveSymlinks,
+			AllowDanglingLinks: in.AllowDanglingSymlinks,
+		}
+		_, err = u.visitFile(ctx, key, info)
 		return err
 	})
 	return nil
 }
 
+// fsCacheKey is a key in u.fsCache.
+//
+// The fsCacheKey is primarily the absolute path, but also configuration of
+// file traversal.
+type fsCacheKey struct {
+	AbsPath            string
+	PreserveSymlinks   bool
+	AllowDanglingLinks bool
```

Comment:
Do you guys have uses where you change these settings? If not we could make these client settings for simplicity? AFAIK we don't use these options, and I can't actually envision use cases where a client will choose different options for different uploads.

Reply:
we don't. done

Reply:
One concern here is that now the Client is not reusable for different parts of the code that need different Upload-specific settings. It is not very nice, not very generic. The prev implementation didn't have this problem, but it had the problem of additional complexity.

I am currently working on file predicate, and this problem seems very similar. I think we need to add a predicate that applies to all `UploadInput`s, but doing it at the client level seems wrong for the same reasons. Therefore I am thinking of adding `UploadOptions`, and change the `Upload` func signature to

```
  func (...) Upload(ctx context, opt UploadOptions, items <- chan *UploadItem) ...
```

The `UploadOptions` would contain `PreserveSymlinks` and `AllowDanglingSymlinks`, therefore allowing reusing the client for different settings, but still not complicated the cache key. I'd probably do it in a separate CL, since this one is reviewed already.

After that, I'd add file predicate to `UploadOptions` too.

How does this sound?

Reply:
I'm not sure I follow how `UploadOptions` would not complicate the cache key. You plan to include it in the cache key right? Do you mean its one struct versus adding a list of different flags and knobs to the cache key explicitly?

Reply:
let's continue in https://github.com/bazelbuild/remote-apis-sdks/pull/314

Reply:
@ramymedhat didn't see your comment earlier. I've responded in https://github.com/bazelbuild/remote-apis-sdks/pull/314#issuecomment-827768249